### PR TITLE
AirfRANS: Pressure-weighted loss (20x) at 4L/256d + golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
BREAKTHROUGH FINDING: Pressure-weighted loss (20x) at 3L/192d achieved val_primary/surface_mse = **0.00435** — 40% better than the 4L/256d baseline (0.007264) and matching the external target (0.0043). The pressure channel dominates the composite MSE but gets equal gradient weight by default, causing gradient misallocation. We hypothesize that combining 20x pressure weighting with the superior 4L/256d + golden config architecture will push BELOW the external target.

## Code Change Required
You need to add `--pressure-loss-weight` support to train.py. This is a small change:

1. Add argument to the parser:
   `parser.add_argument('--pressure-loss-weight', type=float, default=1.0, help='Weight multiplier for pressure channel in loss')`

2. In the loss computation, weight the pressure channel. Find where the MSE loss is computed per-channel and multiply the pressure channel's contribution by this weight. The pressure channel is typically index 2 (after Ux, Uy) in the AirfRANS output. You can also reference nami's implementation on branch `nami/airfrans-pressure-weighted-loss` for guidance.

## Training Command
After implementing the code change:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --pressure-loss-weight 20.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent edward \
  --wandb-name "edward/airfrans-4L256d-pressure20x-golden" \
  --wandb-group "airfrans-pressure-weight"
```

Report val_primary/surface_mse, best epoch, and per-channel breakdown (especially surface_mse_p).

## Baseline
Current best: **0.007264** at epoch 40 (4L/256d, golden config, NO pressure weighting)
Pressure-weight reference: **0.00435** at epoch 241 (3L/192d, lr=3e-4, T_max=10, 20x pressure weight) — W&B run toeli7xw
External target: **0.0043**

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Best Result
| Metric | Value | vs Baseline (0.007264) | vs External Target (0.0043) |
|--------|-------|------------------------|----------------------------|
| **val_primary/surface_mse** | **0.00241** (epoch 153) | **-66.8%** | **-44.0%** |

### Per-channel breakdown (epoch 153)
| Channel | Surface MSE | Volume MSE |
|---------|-------------|------------|
| Ux | 0.0000860 | 0.0547 |
| Uy | 0.00000968 | 0.0386 |
| p (pressure) | 0.00953 | 0.0143 |
| nut | 0.00000925 | 0.0561 |
| **composite** | **0.00241** | **0.0410** |

### Training Summary
- **Epochs:** 200 (killed due to divergence — model diverged catastrophically after epoch 169)
- **Best epoch:** 153 (val_primary/surface_mse = 0.00241)
- **Grad norms:** 50-350 normally, spiking to 15k-19k during late divergence
- **W&B run:** v0pi2h84
- **Peak memory:** ~2.1 GB VRAM (4L/256d with Fourier features)

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --pressure-loss-weight 20.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent edward \
  --wandb-name "edward/airfrans-4L256d-pressure20x-golden" \
  --wandb-group "airfrans-pressure-weight"
```

### What happened

**The hypothesis is strongly confirmed.** Combining 20x pressure-weighted loss with the 4L/256d golden config achieves val_primary/surface_mse = 0.00241 — **44% below the external target** (0.0043) and **66.8% below the baseline** (0.007264). This is also **44.6% better than the 3L/192d reference** (0.00435 at epoch 241 on W&B run toeli7xw).

The training showed three distinct phases:
1. **Epochs 1-30:** Rapid descent through a series of phase transitions (0.44 → 0.04)
2. **Epochs 30-168:** Oscillating descent in the 0.002-0.02 range, with the best troughs progressively deepening (0.018 → 0.012 → 0.007 → 0.006 → 0.004 → 0.0024)
3. **Epochs 169-200:** Catastrophic divergence — val jumped from 0.006 to 0.28 and never recovered. Grad norms exploded from ~50 to 15k-19k.

The 20x pressure weighting creates high peak-to-trough variance (best trough 0.0024, worst peak ~0.11 in epochs 100-168). The deep basins at 0.0024 were achieved at epochs 153 and 155, with the external target beaten at 4 separate epochs (109, 117, 144, 152-153, 155, 162).

The late divergence is the main concern — the aggressive 20x weighting amplifies gradient spikes at cosine LR peaks, eventually causing catastrophic instability.

### Suggested follow-ups

1. **Lower pressure weight (10x)** — may reduce late-training instability while still significantly improving pressure accuracy
2. **Combine with gc=0.5** — current BASELINE best (0.00277) uses gc=0.5; combining lower grad clip with pressure weighting could be synergistic and prevent the late divergence
3. **Reduce LR (e.g., 5e-4)** — would reduce peak-to-trough volatility and may prevent the catastrophic blowup
4. **Checkpoint saving** — the best model at epoch 153 was lost to divergence; saving best checkpoints would preserve winning states
5. **LR warmup or decay** — a decaying max LR schedule could prevent the cosine peaks from becoming destructive in later epochs

**Comment by tcapelle** [comment]
ADVISOR: Good work — pressure-weight 20x improved from 0.007264 to 0.002409 on 4L/256d, confirming the technique works. However, the current baseline is now 0.001479 (PR #2771, **3L/256d** + gc=1.0 + T_max=5). Your 4L/256d result doesn't beat this.

Please re-run pressure weighting at the **winning 3L/256d architecture**. Run a sweep of pressure weights to find the sweet spot:

**Run 1 — pressure-weight 5x + 3L/256d + gc=0.5**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --pressure-loss-weight 5 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-pressure \
  --wandb-name airfrans-3L256d-pw5x-gc05
```

**Run 2 — pressure-weight 10x + 3L/256d + gc=0.5**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --pressure-loss-weight 10 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-pressure \
  --wandb-name airfrans-3L256d-pw10x-gc05
```

**Run 3 — pressure-weight 20x + 3L/256d + gc=0.5**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --pressure-loss-weight 20 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-pressure \
  --wandb-name airfrans-3L256d-pw20x-gc05
```

gc=0.5 should help prevent the divergence you saw at ep169. Baseline to beat: 0.001479.

**Comment by tcapelle** [comment]
STUDENT edward:

## Results — 3L/256d Pressure-Weight Sweep (gc=0.5)

### Summary

| Run | Pressure Weight | Best surface_mse | Best Epoch | Diverged At | vs Baseline (0.001479) | W&B Run |
|-----|----------------|-----------------|------------|-------------|----------------------|---------|
| **pw5x** | **5x** | **0.003261** | 123 | ep154 | +120.5% (2.2x worse) | 66x2i2ba |
| pw10x | 10x | 0.007053 | 63 | ep72 | +376.8% (4.8x worse) | 2ycnxuye |
| **pw20x** | **20x** | **0.003509** | 171 | ep179 | +137.3% (2.4x worse) | h2l7h4fz |

**None of the runs beat the baseline (0.001479).** All three diverged before reaching competitive performance.

### Per-channel breakdown at best epoch

| Channel | pw5x (ep123) | pw10x (ep63) | pw20x (ep171) |
|---------|-------------|-------------|--------------|
| Ux | 0.0000463 | 0.0001147 | 0.0001000 |
| Uy | 0.0000040 | 0.0000146 | 0.0000081 |
| p (pressure) | 0.0129905 | 0.0280657 | 0.0139155 |
| nut | 0.0000040 | 0.0000178 | 0.0000109 |
| **composite** | **0.003261** | **0.007053** | **0.003509** |
| volume_mse | 0.0216 | 0.0869 | 0.0625 |

### Divergence analysis

All three runs eventually diverged to inf grad norms:
- **10x**: Fastest to diverge (ep72). Grad norms were already 150-220 during healthy training, spiked to 4.8M at ep98, inf by ep125.
- **5x**: Diverged at ep154. Had the lowest grad norms during healthy training (~20-40), but instability built gradually. Best val was at ep123.
- **20x**: Last to diverge (ep179). Grad norms were 50-110 during healthy training. Best val at ep171, then grad norms spiked to 19k+ and inf by ep200.

The cosine LR schedule (T_max=5) with gc=0.5 was insufficient to prevent divergence for any pressure weight setting at lr=7e-4. The pressure weighting amplifies gradient magnitudes, and the periodic LR peaks eventually trigger catastrophic instability.

### Commands used

```bash
# pw5x (GPU 0)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --pressure-loss-weight 5 --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent edward \
  --wandb-group airfrans-3L256d-pressure --wandb-name "edward/airfrans-3L256d-pw5x-gc05"

# pw10x (GPU 1) — same but --pressure-loss-weight 10, CUDA_VISIBLE_DEVICES=1
# pw20x (GPU 2) — same but --pressure-loss-weight 20, CUDA_VISIBLE_DEVICES=2
```

### What happened

**Pressure weighting does NOT improve the 3L/256d baseline.** While the 4L/256d experiment (previous results) achieved 0.00241 with 20x weighting, the 3L/256d architecture with gc=0.5 could not reach the same territory. The best result (5x at 0.00326) is 2.2x worse than the baseline (0.001479).

Key observations:
1. **gc=0.5 did delay divergence** compared to the 4L/256d+20x run (which diverged at ep169), but did not prevent it
2. **5x was the most stable** (lowest grad norms, best result) despite 10x diverging first (not 20x as expected)
3. **The baseline's advantage is architectural** — the 3L/256d + gc=1.0 + T_max=5 config at lr=7e-4 without pressure weighting achieves 0.001479, suggesting the unweighted loss landscape is more favorable for this architecture
4. **Pressure weighting creates a gradient magnitude mismatch** — pressure channel gradients dominate, causing instability that gc=0.5 cannot fully control

### Suggested follow-ups

1. **Lower LR with pressure weighting** (e.g. lr=3e-4 or 5e-4) — the main failure mode is gradient instability from LR peaks; a lower base LR might prevent divergence
2. **gc=0.3 or gc=0.2** — more aggressive grad clipping may be needed to stabilize pressure-weighted training
3. **Pressure weight 2x or 3x** — much milder weighting might help without destabilizing
4. **Warmup + decay schedule** — avoid the full LR peaks in early and late training
5. **Separate pressure LR** — decouple the pressure channel learning rate from the rest instead of weighting the loss

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2801 because ADVISOR: Closing — pressure weight experiments at 3L/256d all regressed vs baseline (best 0.003261 vs baseline 0.001236 on AirfRANS). The 3L/256d architecture is too small for AirfRANS and the pressure weighting didn't compensate. The insight about pressure-weighted loss is worth revisiting at the correct architecture scale (2L/256d) if someone picks it up later..
